core/gate.cc

```diff
@@ -42,67 +43,92 @@ namespace bess {
 
 const GateHookCommands GateHook::cmds;
 
-bool GateHookFactory::RegisterGateHook(GateHook::constructor_t constructor,
+bool GateHookBuilder::RegisterGateHook(GateHook::constructor_t constructor,
+                                       const std::string &class_name,
+                                       const std::string &name_template,
+                                       const std::string &help_text,
                                        const GateHookCommands &cmds,
-                                       GateHook::init_func_t init_func,
-                                       const std::string &hook_name,
-                                       const std::string &help_text) {
-  return all_gate_hook_factories_holder()
-      .emplace(std::piecewise_construct, std::forward_as_tuple(hook_name),
-               std::forward_as_tuple(constructor, cmds, init_func, hook_name,
-                 help_text))
+                                       GateHook::init_func_t init_func) {
+  return all_gate_hook_builders_holder()
+      .emplace(std::piecewise_construct, std::forward_as_tuple(class_name),
+               std::forward_as_tuple(constructor, class_name, name_template,
+                 help_text, cmds, init_func))
       .second;
 }
 
-std::map<std::string, GateHookFactory>
-    &GateHookFactory::all_gate_hook_factories_holder(bool reset) {
-  // Maps from hook names to hook factories. Tracks all hooks (via their
-  // GateHookFactorys).
-  static std::map<std::string, GateHookFactory> all_gate_hook_factories;
+std::map<std::string, GateHookBuilder>
+    &GateHookBuilder::all_gate_hook_builders_holder(bool reset) {
+  // Maps from hook names to hook builders. Tracks all hooks (via their
+  // GateHookBuilders).
+  static std::map<std::string, GateHookBuilder> all_gate_hook_builders;
 
   if (reset) {
-    all_gate_hook_factories.clear();
+    all_gate_hook_builders.clear();
   }
 
-  return all_gate_hook_factories;
+  return all_gate_hook_builders;
 }
 
-const std::map<std::string, GateHookFactory>
-    &GateHookFactory::all_gate_hook_factories() {
-  return all_gate_hook_factories_holder();
+const std::map<std::string, GateHookBuilder>
+    &GateHookBuilder::all_gate_hook_builders() {
+  return all_gate_hook_builders_holder();
 }
 
-// Creates new gate hook from the given factory, initializes it,
+const std::string Gate::GenerateDefaultName(const GateHookBuilder *builder,
+    Gate *gate) {
+  const std::string &name_template = builder->name_template();
+
+  for (int i = 0;; i++) {
+    const std::vector<GateHook *> hooks = gate->hooks();
+    std::string name = name_template + std::to_string(i);
+    auto it = std::find_if(hooks.begin(), hooks.end(),
```

Comment:
Just a matter of style, but you may want to simplify here like this (not tested). I personally find using lambdas more intuitive than using functors...
```
if (std::none_of(hooks.begin(), hooks.end(), [&](const GateHook *hook){ return name == hook->name(); }) {
  return name;
}
```

There are some `find_if`-then-check or manual `for` loops on many occasions of the code, which can be also updated.